syntax = "proto3";
//MapComplementaryInformationEvent
message MapComplementaryInformationEvent {
    uint32 subarea_id = 1;
    uint64 map_id = 2;
    repeated House houses = 3;
    repeated ActorPositionInformation actors_position_information = 4;
    repeated InteractiveElement interactive_elements = 5;
    repeated StatedElement stated_elements = 6;
    bool has_aggressive_monsters = 9;

    message InteractiveElement {
        int32 element_id = 1;
        int32 element_type_id = 2;
        repeated InteractiveElementSkill enabled_skills = 3;
        repeated InteractiveElementSkill disabled_skills = 4;
        bool on_current_map = 5;
        int32 age_bonus = 6;

        message InteractiveElementSkill {
            int32 skill_id = 1;
            int32 skill_instance_uid = 2;
            int32 name_id = 3;
        }
    }

    message StatedElement {
        int32 element_id = 1;
        int32 cell_id = 2;
        int32 state = 3;
        bool on_current_map = 4;
    }

    message ActorPositionInformation {
        int64 actor_id = 1;
        EntityDisposition disposition = 2;
        ActorInformation actor_information = 3;

        message ActorInformation {
            EntityLook look = 1;
            oneof information {
                RolePlayActor role_play_actor = 2;
                Fighter fighter = 3;
            }
            enum InformationOneofCase {
                None = 0;
                RolePlayActor = 2;
                Fighter = 3;
            }
        }
        enum ActorOneofCase {
            NONE = 0;
            NAMED_ACTOR = 4;
            TAX_COLLECTOR_ACTOR = 5;
            MONSTER_GROUP_ACTOR = 6;
            NPC_ACTOR = 7;
            PRISM_ACTOR = 8;
            PORTAL_ACTOR = 9;
            TREASURE_HUNT_NPC_ID = 10;
        }

        message RolePlayActor {
            oneof actor {
                NamedActor named_actor = 4;
                MonsterGroupActor monster_group_actor = 6;
                int32 treasure_hunt_npc_id = 10;
            }

            message NamedActor {
                string name = 1;

                oneof actor {
                    Humanoid humanoid_information = 2;
                    Mount mount_information = 3;
                }

                message Humanoid {
                    repeated HumanOption options = 3;
                    int64 account_id = 4;
                    MutantInformation mutant_information = 6;

                    message HumanOption {
                        oneof option {
                            Followers followers = 3;
                            AllianceInformation alliance_information = 5;
                            EmoteOption emote_option = 6;
                            TitleOption title_option = 7;
                            OrnamentOption ornament_option = 8;
                            ObjectUseOption object_use_option = 9;
                            SkillUseOption skill_use_option = 10;
                            float speed_multiplier = 11;
                        }

                        message Followers {
                            repeated IndexedEntityLook following_characters_looks = 1;
                        }

                        message IndexedEntityLook {
                            EntityLook look = 1;
                            int32 index = 2;
                        }

                        message AllianceInformation {
                            // Define fields as per your needs
                        }

                        message EmoteOption {
                            int32 emote_id = 1;
                            string emote_start_time = 2;
                        }

                        message TitleOption {
                            int32 title_id = 1;
                            string title_parameter = 2;
                        }

                        message OrnamentOption {
                            int32 ornament_id = 1;
                            int32 player_level = 2;
                            int32 league_id = 3;
                            int32 ladder_position = 4;
                        }

                        message ObjectUseOption {
                            int32 delayed_action_type = 1;
                            string end_delay = 2;
                            int32 object_gid = 3;
                        }

                        message SkillUseOption {
                            int32 element_id = 1;
                            int32 skill_id = 2;
                            string skill_end_time = 3;
                        }
                    }

                    message MutantInformation {
                        int32 monster_id = 1;
                        int32 power_level = 2;
                    }
                }

                message Mount {
                    string owner_name = 1;
                    int32 level = 2;
                }
            }


            message MonsterGroupActor {
                MonsterGroupStaticInformation identification = 1;
                int32 loot_share = 2;
                message WaveInformation {
                    int32 wave_count = 1;
                    repeated MonsterGroupStaticInformation alternatives = 2;
                }

                message MonsterGroupStaticInformation {
                    MonsterInGroupInformation main_creature = 1;
                    repeated MonsterInGroupInformation underlings = 2;

                    message MonsterInGroupInformation {
                        int32 gid = 1;
                        int32 grade = 2;
                        int32 level = 3;
                    }

                }
            }

        }


        message EntityLook {
            string description = 1;
        }


        message Fighter {
            int32 strength = 1;
        }
    }

    enum EntityDisposition {
        UNKNOWN = 0;
        ACTIVE = 1;
        INACTIVE = 2;
    }

    message House {
        int32 house_id = 1;
        int32 model_id = 2;

        oneof information {
            HouseAccount house_account = 3;
            HouseOnMap house_on_map = 4;
            HouseInside house_inside = 5;
            HouseGuild house_guild = 6;
        }
    }

    message HouseAccount {
        HouseInstance house_information = 1;
        MapExtendedCoordinates coordinates = 2;
    }

    message HouseOnMap {
        repeated int32 doors_on_map = 1;
        repeated HouseInstance houses_information = 2;
    }

    message HouseInside {
        HouseInstance house_information = 1;
        MapCoordinates coordinates = 2;
    }

    message HouseGuild {
        HouseInstance house_information = 1;
        MapExtendedCoordinates coordinates = 2;
        repeated int32 skills_id = 3;
        int64 guild_share_params = 4;
    }
    message MapExtendedCoordinates {
    int32 world_x = 1;
    int32 world_y = 2;
    int64 map_id = 3;
    int32 sub_area_id = 4;
}
    message MapCoordinates {
    int32 world_x = 1;

    int32 world_y = 2;
}
    message HouseInstance {
    int32 instance_id = 1;
    bool second_hand = 2;
    bool is_locked = 3;
    AccountTag account_tag = 4;
    bool has_owner = 5;
    int64 price = 6;
    bool is_sale_locked = 7;
    bool is_admin_locked = 8;
//    GuildInformation guild_information = 9;
    int32 rooms_count = 10;
    int32 chests_count = 11;
    repeated int32 skills_id = 12;
}
    message AccountTag {
    string nickname = 1;

    string tag = 2;
}
}

//TeleportDestinationsEvent
message TeleportDestinationsEvent {
    Teleporter type = 1;
    repeated TeleportDestination destinations = 2;
    int64 spawnMapId = 3;
}
message TeleportDestination {
  enum ComplementaryInformationOneofCase {
    NONE = 0;
    ANOMALY = 6;
  }
  message Anomaly {
    int32 bonusPourcentage = 1;
    int64 remainingTime = 2;
  }
  Teleporter type = 1;
  int64 mapId = 2;
  int32 subareaId = 3;
  int32 level = 4;
  int32 cost = 5;
  oneof complementaryInformation {
    Anomaly anomaly = 6;
  }
}
enum Teleporter {
    TELEPORTER_ZAAP = 0;
    TELEPORTER_SUBWAY = 1;
    TELEPORTER_PRISM = 2;
    TELEPORTER_HAVEN_BAG = 3;
    TELEPORTER_ANOMALY = 4;
}
//InteractiveUseEndedEvent
message InteractiveUseEndedEvent {
    int32 element_id = 1;
    int32 skill_id = 2;
}
//FightFighterShowEvent
message FightFighterShowEvent {
  bool static_pose = 2;
}
//FightEndEvent
message FightEndEvent {
    int32 duration = 1;
    int32 rewardRate = 2;
}
//FightNewRoundEvent
message FightNewRoundEvent {
    int32 round_number = 1;
}
//MapMovementEvent
message SequenceStartEvent {
    int64 author_id = 2;
}
message MapMovementEvent {
    repeated int32 cells = 1;
    int32 direction = 2;
    int64 character_id = 3;
    bool cautious = 4;
}
//InventoryWeightEvent
message InventoryWeightEvent {

  int32 inventoryWeight = 1;
  int32 weightMax = 2;
}